### PR TITLE
Reverse proxy

### PR DESCRIPTION
**Pull request by neerajshukla1911**
Added reverse proxy server

Jira Ticket: <https://browserstack.atlassian.net/browse/ACE-1056>

### CODE REVIEW

**Review by rchougule — CHANGES_REQUESTED** [review]
Majority of the code in `reverseProxy.js` is common with `proxy.js`. Possible to take out the common part in a base class/file?

File: config/constants.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 module.exports.VERSION = '1.0.0';
+module.exports.HUB_HOST = 'hub-cloud.browserstack.com';
 module.exports.HUB_STATUS_URL = 'http://hub-cloud.browserstack.com/wd/hub/status';
```
Comment (Archish27) [inline config/constants.js]: can we utilise constant `HUB_HOST` in `HUB_STATUS_URL`?

File: test/reverseProxy.test.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
+var constants = require('../config/constants');
+var RdGlobalConfig = constants.RdGlobalConfig;
+var nock = require('nock');
+var reverseProxy = require('../src/reverseProxy');
+var http = require('http');
+var assert = require('chai').assert;
+var testHelper = require('./testHelper');
+
+var REVERSE_PROXY_DOMAIN = `http://127.0.0.1:${RdGlobalConfig.RD_HANDLER_REVERSE_PROXY_PORT}`;
```
Comment (Archish27) [inline test/reverseProxy.test.js]: Constant?

File: test/reverseProxy.test.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+var HUB_STATUS_URI_PATH = "/wd/hub/status";
```
Comment (Archish27) [inline test/reverseProxy.test.js]: Can we utilize this from constants? Or create a new constant?

File: src/requestLib.js [hunk 4]
```diff
@@ -15,38 +14,74 @@ var RequestLib = {
...
+          });
+
+          response.on('end', function () {
+            responseToSend.data = Buffer.concat(responseToSend.data).toString();
+            resolve(responseToSend);
+          });
+
+          response.on('error', function (err) {
+            reject({
+              message: err,
```
Comment (Archish27) [inline src/requestLib.js]: Can this be moved to function? Duplicates with `https`?

File: src/commandLine.js [hunk 5]
```diff
@@ -78,7 +78,7 @@ var CommandLineManager = {
         argv.splice(index, 1);
       }
     }
-
+
```
Comment (rchougule) [inline src/commandLine.js:81]: whitespace...

Comment (rchougule) [reply src/commandLine.js:81]: and at other places as well..

File: src/requestLib.js [hunk 6]
```diff
@@ -13,32 +12,36 @@ var RequestLib = {
...
    */
-  _makeRequest: function (params, clientRequest, retries) {
+  _makeRequest: function (schemeObj, params, clientRequest, retries) {
     return new Promise(function (resolve, reject) {
-      var requestOptions = Object.assign({}, params.furtherRequestOptions, {
+      var requestOptions = {
+        headers:{}
+      };
+      // Initialize the request to be fired on behalf of the client
+      var request = null;
```
Comment (rchougule) [inline src/requestLib.js]: can be moved back to line 29...

File: src/requestLib.js [hunk 7]
```diff
@@ -13,32 +12,36 @@ var RequestLib = {
...
+      var keepAliveAgent = null;
+      keepAliveAgent = new schemeObj.Agent({keepAlive: true});
```
Comment (rchougule) [inline src/requestLib.js]: no need to create the agent again and again... can be selected conditionally wherever required...

File: src/requestLib.js [hunk 8]
```diff
@@ -13,32 +12,36 @@ var RequestLib = {
    * @param {http.IncomingMessage} clientRequest 
    * @param {Number} retries 
    */
-  _makeRequest: function (params, clientRequest, retries) {
+  _makeRequest: function (schemeObj, params, clientRequest, retries) {
     return new Promise(function (resolve, reject) {
-      var requestOptions = Object.assign({}, params.furtherRequestOptions, {
+      var requestOptions = {
+        headers:{}
+      };
```
Comment (rchougule) [inline src/requestLib.js]: why do we need to do this?...

File: src/server.js [hunk 9]
```diff
@@ -29,52 +28,44 @@ var RdHandler = {
       port: null,
       path: null
     };
-
+
```
Comment (rchougule) [inline src/server.js:31]: Please use the eslint config. Can be setup in the respective IDE for these...

File: src/server.js [hunk 10]
```diff
@@ -29,52 +28,44 @@ var RdHandler = {
...
-        var requestOptions = Object.assign({}, RdHandler._reqObjTemplate);
-        requestOptions.host = parsedClientUrl.hostname;
-        requestOptions.port = parsedClientUrl.port || 80;
+    }   
+    var requestOptions = Object.assign({}, RdHandler._reqObjTemplate);
+
+
+    /**
+     * Sets the internal method to generate request options
+     * doesn't exists
```
Comment (rchougule) [inline src/server.js]: comment seems broken...

**Comment by neerajshukla1911** [comment]
Raised another PR with https support

**Pull request closed without merging**